### DISCUSSION TITLE
[Potential issue] `Mapped` as asymmetric descriptor

### DISCUSSION THREAD

**Discussion by AlexanderPodorov** [sqlalchemy/sqlalchemy#12056] [Usage Questions] [answered]
Is it intended behavior?

Here's the relevant type definition from the sqlalchemy code. Note that the `__set__` accepts a value of type `SQLCoreOperations[_T_co] | _T_co` but returns a value of type `_T_co`.

```python
class Mapped:
    ...
    def __get__(self, instance: object, owner: Any) -> _T_co: ...
    def __set__(self, instance: Any, value: SQLCoreOperations[_T_co] | _T_co) -> None: ...
```

See the related pyright issue for reference: https://github.com/microsoft/pyright/issues/9372

**Comment by CaselIT** [comment]
Hi,

That's not the correct annotation for mapped. Get may return both versions depending if it's called at class or instance level:
https://github.com/sqlalchemy/sqlalchemy/blob/21ea01eebe0350ad1185c7288dca61f363ebd2fe/lib/sqlalchemy/orm/base.py#L810-L827

**Comment by AlexanderPodorov** [comment]
Hi, thanks, should I create an issue for that?

**Reply by CaselIT** [reply]
I'm not sure what issue you are referring towards

**Reply by AlexanderPodorov** [reply]
I mean to create an issue to fix the type annotations for `Mapped`

**Reply by CaselIT** [reply]
I'm unaware of any issue with the type annotations for mapped

**Reply by AlexanderPodorov** [reply]
Oh, yeah, I just copied the code snippet from pyright's maintainer's response. Some parts were omitted here.
He noted that the following methods are "asymmetric", therefore it causes a typing issue. It would be great if you could take a look at the original issue I created for `pyright` https://github.com/microsoft/pyright/issues/9372, for the full context. Thank you!
![image](https://github.com/user-attachments/assets/cf76974a-c2e1-43e5-a915-38640de9097f)

**Reply by CaselIT** [reply, marked as answer]
that's the correct typing for Mapped.
